feat: Postgres Prayers

tomthornton opened the pull request:
In order to run, import `PrayerRequest` from `apollos-dataconnector-postgres` and remove the import from `apollos-dataconnector-rock` in the `index.postgres.js` file of your ApolloChurchApi

vinnyjth commented:
At a glance, looking great!

Remember tests, we’ll need them before we merge :) try to get 100% test coverage over the dataSource methods as a goal

> On Aug 19, 2021, at 4:52 PM, codecov[bot] ***@***.***> wrote:
> 
> ﻿
> Codecov Report
> 
> Merging #2138 (2f2b523) into master (1b92606) will decrease coverage by 0.35%.
> The diff coverage is 12.96%.
> 
> 
> 
> @@            Coverage Diff             @@
> ##           master    #2138      +/-   ##
> ==========================================
> - Coverage   75.77%   75.41%   -0.36%     
> ==========================================
>   Files         585      590       +5     
>   Lines        9072     9123      +51     
>   Branches     2225     2231       +6     
> ==========================================
> + Hits         6874     6880       +6     
> - Misses       1949     1990      +41     
> - Partials      249      253       +4     
> Impacted Files	Coverage Δ	
> ...-connector-postgres/src/action-algorithms/index.js	1.23% <0.00%> (+0.01%)	⬆️
> ...-data-connector-postgres/src/prayers/dataSource.js	0.00% <0.00%> (ø)	
> ...os-data-connector-postgres/src/prayers/resolver.js	0.00% <0.00%> (ø)	
> ...tgres/src/prayers/migrations/001_create_prayers.js	28.57% <28.57%> (ø)	
> ...ollos-data-connector-postgres/src/prayers/model.js	40.00% <40.00%> (ø)	
> ...connector-postgres/src/prayers/migrations/index.js	100.00% <100.00%> (ø)	
> ...ata-connector-rock/src/interactions/data-source.js	80.55% <100.00%> (+0.27%)	⬆️
> ... and 2 more		
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

tomthornton commented:
@vinnyjth tests added!

vinnyjth commented on packages/apollos-data-connector-postgres/src/prayers/__tests__/dataSource.test.js:
```diff
@@ -0,0 +1,120 @@
+    await prayer1.reload();
+
+    expect(prayer1.usersPrayed).toContain(currentPerson.id);
+  });
```
Actually I take that back! We don't have flagging :)We probably should though! Since Rock supports it. We can add a field in a follow up PR.

vinnyjth commented on packages/apollos-data-connector-rock/src/interactions/data-source.js:
```diff
@@ -198,14 +198,19 @@ export default class Interactions extends RockApolloDataSource {
+
+    // Currently it throws an error for a postgres created prayer, this will be fixed once they are shoveled back
+    // into Rock or on command during the prayer creation
+    if (action !== 'PRAY') {
```
We'll likely have to be a bit more careful here :/ if we have this `action` skipping the `interaction` all of the rock folks are going to see their `Prayer` interactions drop off. Maybe we could do something (even if it's hacky) to figure out if it's a Rock or Postgres ID and then skip the `post` then? or even wrap the `post` in a try/catch

tomthornton replied:
Yeah I figured this would be the case. I like the try/catch fix for now. Once they are shoveled back into rock it should still work.

vinnyjth commented on packages/apollos-data-connector-postgres/src/prayers/dataSource.js:
```diff
@@ -0,0 +1,115 @@
+    return sortedPrayerRequests;
+  }
+
+  async getPrayerById(prayerId) {
```
You might be able to get away without this method by using the inherited `getFromId` method.

vinnyjth commented on packages/apollos-data-connector-postgres/src/prayers/dataSource.js:
```diff
@@ -0,0 +1,115 @@
+    });
+  }
+
+  async incrementPrayed(prayerId) {
```
Is this working? I believe the function signature here (coming from the additional interactions) is an object of the form `{ id, __type, action }`

tomthornton replied:
Yep it's working. Line 18 of the interactions data source in Rock takes the id out and passes it to the incrementPrayer function.

vinnyjth replied:
Ah, thanks I missed that. 👍

vinnyjth reviewed (commented):
Here's my initial round of feedback @tomthornton  :D Looking pretty good so far, I think the only major thing we need to figure out is how to make the `prayed_for` array not a JSONB field :) 

I'm going to test in the sim, and then I might have more feedback.

vinnyjth commented on packages/apollos-data-connector-postgres/src/prayers/migrations/001_create_prayers.js:
```diff
@@ -0,0 +1,55 @@
+import { Sequelize, DataTypes } from 'sequelize';
+
+async function up({ context: queryInterface }) {
+  await queryInterface.createTable('prayer', {
```
What do you think about naming this table `prayer_request`?

tomthornton replied:
Sure that makes sense. And the join table `user_prayer` or just `prayer`?

vinnyjth commented on packages/apollos-data-connector-postgres/src/prayers/dataSource.js:
```diff
@@ -0,0 +1,115 @@
+    return newPrayer[0];
+  }
+
+  async byDailyPrayerFeed({ numberDaysSincePrayer = 3, limit = 10 }) {
```
I believe this `method` takes an optional `personId` argument that, when provided, returns _only_ prayers from that person. We'll need to implement that functionality to support this "My Prayers" feature. (Right now it's showing the same content as the "Pray for others" feature. 

![image](https://user-images.githubusercontent.com/1637694/130455436-bd2c380f-91fa-4cd4-a274-8334ccd4472e.png)

vinnyjth reviewed (changes requested):
I think that's all the feedback I have :D thanks for your hard work Tom! PR is looking really good

tomthornton commented:
@vinnyjth Thank you! I'll be working on those changes

camrun91 commented:
While you all are working on this @vinnyjth @tomthornton it would be awesome to have the ability to mark your own prayers as answered with an answer text Rock supports this currently. Also would be neat if you prayed for a prayer to get notified if it is marked answered. 

If I need to move this to discussions I can do that.

tomthornton commented:
@camrun91 yeah that sounds like a cool feature! We'll probably have to wait to implement that until after we've connected things back to rock.

tomthornton commented:
- Added `getCurrentPerson` to the People datasource... not sure if this is already accessible somewhere else
- Added joint table `person_prayed_for` and some foreign key constraints to help with deleting
- Updated tests

vinnyjth commented:
@tomthornton I think there's a few more review comments that might require changes. Do you want to take a look at those before I re-review?

tomthornton commented:
@vinnyjth ready for review now. I added the `personId` argument to `dailyPrayerFeed` and wrapped the interaction rock call with a try/catch.

vinnyjth commented on packages/apollos-data-connector-postgres/src/prayers/dataSource.js:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  async isPrayedFor({ prayerRequest, currentPersonId }) {
```
Have you considered using an `includes: [...]` to eager load this relationship? Might help reduce the number of queries significantly in a longer list of prayers :)

vinnyjth replied:
If you add it as a default scope (see `content_items` with cover images`) you can know that the relationship will always be present

tomthornton replied:
Yep I was thinking the same thing, didn't know you could do that. I'll try it out.

vinnyjth commented on packages/apollos-data-connector-postgres/src/people/dataSource.js:
```diff
@@ -141,7 +141,7 @@ export default class Person extends PostgresDataSource {
-  async getCurrentPersonId() {
+  async getCurrentPerson() {
     if (this.context.currentPostgresPerson) {
       return this.context.currentPostgresPerson.id;
```
Returning `id` here probably will result in a bug :)

vinnyjth commented on packages/apollos-data-connector-postgres/src/prayers/dataSource.js:
```diff
@@ -0,0 +1,121 @@
+  // eslint-disable-next-line class-methods-use-this
+  async isPrayedFor({ prayerRequest, currentPersonId }) {
+    return prayerRequest
+      .getPrayedUsers({
```
@tomthornton I think calling `.getPrayedUsers` here will re-run the query, rather than using the eager-loading that you added earlier with the `defaultScope`, Is that intentional?

vinnyjth replied:
Actually I can't prove this! So I think we're good.

vinnyjth replied:
😅 I realized I can't prove it, because this code isn't getting called anymore. You should be safe to remove it

vinnyjth commented on packages/apollos-data-connector-postgres/src/prayers/migrations/001_create_prayers.js:
```diff
@@ -0,0 +1,55 @@
+        key: 'id',
+      },
+    },
+    users_prayed: {
```
@tomthornton You can delete `users_prayed` from the migration, I think :) if we had deployed this code, we would have to do another migration, but in this instance it's totally safe to simply remove it from the initial migration.

vinnyjth commented:
@tomthornton 

Just three things to resolve and then we can merge this! Works well in the sim :) 

1. Dead code in the initial migration
2. Dead code in the dataSource `isPrayedFor` method
3. Tiny bug in `getCurrentPerson` logic

tomthornton commented:
@vinnyjth what is the dead code in the initial migration?
